File: src/queue.cpp
#include "queue.h"
Queue::Queue(){
    front = nullptr;
    rear = nullptr;
}

Queue::~Queue(){
    while (front){
        QueueNode* cur = front;
        front = front -> next;
        delete cur;
    }

}

void Queue::pushQ(string value) { //добавляется в конец
    QueueNode* newNode = new QueueNode{value, nullptr}; //создаем новый узел
    if (rear == nullptr) { //если очередь пустая 
        front = rear = newNode; 
        return;
    }
    rear -> next = newNode; //добавляем в конец узел
    rear = newNode; //сдвигаем rear
}

string Queue::popQ() { //удаляется первый элемент
    if (front == nullptr) {
        return "";
    }
    QueueNode* temp = front;//
    string value = temp -> data;
    front = front -> next; //сдвигаем front
    if (front == nullptr)  // если очередь опустела
        rear = nullptr;
    delete temp;
    return value; //выводим удаленный элемент
}

bool Queue::isEmptyQ() {
    return front == nullptr;
}

// Вывод очереди
void Queue::printQ() {
    if (isEmptyQ()) {
        cout << "Очередь пуста" << endl;
        return;
    }
    QueueNode* current = front;
    cout << "Очередь (слева направо): ";
    while (current != nullptr) { //проходим от front до rear
        cout << current -> data << " ";
        current = current -> next;
    }
    cout << endl;
}

// Загрузка очереди из файла
void Queue::loadQ(const string& filename) {
    ifstream fin(filename);
    if (!fin) return;
    string value;
    while (fin >> value) pushQ(value);
    fin.close();
}

// Сохранение очереди в файл
void Queue::saveQ(const string& filename) {
    ofstream fout(filename);
    QueueNode* current = front;
    while (current != nullptr) {
        fout << current -> data << "\n";
        current = current -> next;
    }
    fout.close();
}

File: laba3_go/queue.go
package classes

import (
  "fmt"
)
// private node

type queueNode struct {
  data string
  next *queueNode
}
// Queue (PUBLIC)

type Queue struct {
  front *queueNode // private
  rear  *queueNode // private
}
=============

func NewQueue() *Queue { // PUBLIC
  return &Queue{
    front: nil,
    rear:  nil,
  }
}

// Public methods

// добавление в конец очереди
func (q *Queue) Push(value string) {
  newNode := &queueNode{
    data: value,
    next: nil,
  }

  if q.rear == nil { // очередь пустая
    q.front = newNode
    q.rear = newNode
    return
  }

  q.rear.next = newNode
  q.rear = newNode
}

// удаление первого элемента
func (q *Queue) Pop() string {
  if q.front == nil {
    return ""
  }

  value := q.front.data
  q.front = q.front.next

  if q.front == nil { // очередь стала пустой
    q.rear = nil
  }

  return value
}

func (q *Queue) IsEmpty() bool {
  return q.front == nil
}

func (q *Queue) Print() {
  if q.IsEmpty() {
    fmt.Println("Очередь пуста")
    return
  }

  fmt.Print("Очередь (слева направо): ")
  cur := q.front
  for cur != nil {
    fmt.Print(cur.data, " ")
    cur = cur.next
  }
  fmt.Println()
}
